File: app/src/main/java/org/ametro/utils/StringUtils.java
package org.ametro.utils;

import java.text.Collator;

public class StringUtils {

    private static final Collator COLLATOR = Collator.getInstance();

    private static final String[] SI_UNITS = {"k","M","G","T","P","E"};
    private static final String[] COMPUTING_UNITS = {"Ki","Mi","Gi","Ti","Pi","Ei"};

    static{
        COLLATOR.setStrength(Collator.PRIMARY);
    }

    public static String humanReadableByteCount(long bytes, boolean si) {
        int unit = si ? 1000 : 1024;
        if (bytes < unit) return bytes + " B";
        int exp = (int) (Math.log(bytes) / Math.log(unit));
        String pre = (si ? SI_UNITS : COMPUTING_UNITS)[exp-1];
        return String.format("%.1f %sB", bytes / Math.pow(unit, exp), pre);
    }

    public static boolean startsWithoutDiacritics(String text, String prefix){
        final int textLength = text.length();
        final int prefixLength = prefix.length();
        if(textLength<prefixLength){
            return false;
        }
        String textPrefix = text.substring(0, prefixLength);
        return COLLATOR.compare(textPrefix, prefix) == 0;
    }

    public static boolean isNullOrEmpty(String value) {
        return value == null || "".equals(value.trim());
    }


    public static String humanReadableTime(int totalSeconds) {
        int seconds = totalSeconds % 60;
        int minutes = (totalSeconds / 60) % 60;
        int hours = totalSeconds / 60 / 60;

        if(hours>0){
            return String.format("%02d:%02d:%02d", hours, minutes, seconds);
        }

        return String.format("%02d:%02d", minutes, seconds);
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/utils/StringUtils.kt
package io.github.romangolovanov.apps.ametro.utils

import java.text.Collator

object StringUtils {

    private val COLLATOR: Collator = Collator.getInstance().also {
        it.strength = Collator.PRIMARY
    }

    private val SI_UNITS = arrayOf("k", "M", "G", "T", "P", "E")
    private val COMPUTING_UNITS = arrayOf("Ki", "Mi", "Gi", "Ti", "Pi", "Ei")

    @JvmStatic
    fun humanReadableByteCount(bytes: Long, si: Boolean): String {
        val unit = if (si) 1000 else 1024
        if (bytes < unit) return "$bytes B"
        val exp = (Math.log(bytes.toDouble()) / Math.log(unit.toDouble())).toInt()
        val pre = (if (si) SI_UNITS else COMPUTING_UNITS)[exp - 1]
        return String.format("%.1f %sB", bytes / Math.pow(unit.toDouble(), exp.toDouble()), pre)
    }

    @JvmStatic
    fun startsWithoutDiacritics(text: String, prefix: String): Boolean {
        val textLength = text.length
        val prefixLength = prefix.length
        if (textLength < prefixLength) {
            return false
        }
        val textPrefix = text.substring(0, prefixLength)
        return COLLATOR.compare(textPrefix, prefix) == 0
    }

    @JvmStatic
    fun isNullOrEmpty(value: String?): Boolean {
        return value == null || value.trim().isEmpty()
    }

    @JvmStatic
    fun humanReadableTime(totalSeconds: Int): String {
        val seconds = totalSeconds % 60
        val minutes = (totalSeconds / 60) % 60
        val hours = totalSeconds / 60 / 60

        if (hours > 0) {
            return String.format("%02d:%02d:%02d", hours, minutes, seconds)
        }

        return String.format("%02d:%02d", minutes, seconds)
    }
}
